// samchon/templates/distributed/derived/DistributedClientArrayMediator.js
"use strict";
/// <reference path="../DistributedSystemArrayMediator.ts" />
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var DistributedSystemArrayMediator_1 = require("../DistributedSystemArrayMediator");
/**
 * Mediator of Distributed Processing System, a server accepting slave clients.
 *
 * The {@link DistributedClientArrayMediator} is an abstract class, derived from {@link DistributedSystemArrayMediator}
 * class, opening a server accepting {@link DistributedSystem distributed clients} as a **master**.
 *
 * Extends this {@link DistributedClientArrayMediator}, overrides {@link createServerBase createServerBase()} to
 * determine which protocol to follow and {@link createExternalClient createExternalClient()} creating child
 * {@link DistributedSystem} object. After the extending and overridings, open this server using the
 * {@link open open()} method.
 *
 * #### [Inherited] {@link DistributedSystemArrayMediator}
 * @copydoc DistributedSystemArrayMediator
 */
var DistributedClientArrayMediator = /** @class */ (function (_super) {
    __extends(DistributedClientArrayMediator, _super);
    /* =========================================================
        CONSTRUCTORS
            - MEMBER
            - FACTORY METHOD FOR CHILDREN
    ============================================================
        MEMBER
    --------------------------------------------------------- */
    /**
     * Default Constructor.
     */
    function DistributedClientArrayMediator() {
        return _super.call(this) || this;
    }
    /* ---------------------------------------------------------
        FACTORY METHOD FOR CHILDREN
    --------------------------------------------------------- */
    /**
     * Add a newly connected remote client.
     *
     * When a {@link IClientDriver remote client} connects to this *master server of distributed processing system*,
     * then this {@link DistributedClientArrayMediator} creates a child {@link Distributed distributed client} object
     * through the {@link createExternalClient createExternalClient()} method.
     *
     * @param driver A communicator for external client.
     */
    DistributedClientArrayMediator.prototype.addClient = function (driver) {
        var system = this.createExternalClient(driver);
        if (system == null)
            return;
        this.push_back(system);
    };
    /**
     * (Deprecated) Factory method creating child object.
     *
     * The method {@link createChild createChild()} is deprecated. Don't use and override this.
     *
     * Note that, the {@link DistributedClientArrayMediator} is a server accepting {@link DistributedSystem distributed
     * clients} as a master. There's no way to creating the {@link DistributedSystem distributed clients} in advance
     * before opening the server.
     *
     * @param xml An {@link XML} object represents the child {@link DistributedSystem} object.
     * @return null
     */
    DistributedClientArrayMediator.prototype.createChild = function (xml) { return null; };
    /* ---------------------------------------------------------
        SERVER's METHOD
    --------------------------------------------------------- */
    /**
     * @inheritdoc
     */
    DistributedClientArrayMediator.prototype.open = function (port) {
        this.server_base_ = this.createServerBase();
        if (this.server_base_ == null)
            return;
        this.server_base_.open(port);
    };
    /**
     * @inheritdoc
     */
    DistributedClientArrayMediator.prototype.close = function () {
        if (this.server_base_ == null)
            return;
        this.server_base_.close();
        this.clear();
    };
    return DistributedClientArrayMediator;
}(DistributedSystemArrayMediator_1.DistributedSystemArrayMediator));
exports.DistributedClientArrayMediator = DistributedClientArrayMediator;
