// ecol/lib/linear/DequeCollection.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Deque_1 = require("tstl/container/Deque");
var CollectionEvent_1 = require("../basic/CollectionEvent");
var EventDispatcher_1 = require("../basic/EventDispatcher");
var DequeCollection = /** @class */ (function (_super) {
    __extends(DequeCollection, _super);
    function DequeCollection() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * @hidden
         */
        _this.dispatcher_ = new EventDispatcher_1.EventDispatcher();
        return _this;
    }
    /* ---------------------------------------------------------
        CONSTRUCTORS
    --------------------------------------------------------- */
    // using super.constructor;
    DequeCollection.prototype.clear = function () {
        var first = this.begin();
        var last = this.end();
        this._Notify_erase(first, last);
        _super.prototype.clear.call(this);
    };
    /* =========================================================
        ELEMENTS I/O
            - INSERT
            - ERASE
            - REFRESH
    ============================================================
        INSERT
    --------------------------------------------------------- */
    /**
     * @inheritdoc
     */
    DequeCollection.prototype.push_front = function (val) {
        _super.prototype.push_front.call(this, val);
        this._Notify_insert(this.begin(), this.begin().next());
    };
    /**
     * @inheritdoc
     */
    DequeCollection.prototype.push_back = function (val) {
        _super.prototype.push.call(this, val);
        this._Notify_insert(this.end().prev(), this.end());
    };
    /**
     * @hidden
     */
    DequeCollection.prototype._Insert_by_range = function (pos, first, last) {
        var n = this.size();
        var ret = _super.prototype._Insert_by_range.call(this, pos, first, last);
        n = this.size() - n;
        this._Notify_insert(ret, ret.advance(n));
        return ret;
    };
    /* ---------------------------------------------------------
        ERASE
    --------------------------------------------------------- */
    /**
     * @inheritdoc
     */
    DequeCollection.prototype.pop_front = function () {
        this._Notify_erase(this.begin(), this.begin().next());
        _super.prototype.pop_front.call(this);
    };
    /**
     * @inheritdoc
     */
    DequeCollection.prototype.pop_back = function () {
        this._Notify_erase(this.end().prev(), this.end());
        _super.prototype.pop_back.call(this);
    };
    /**
     * @hidden
     */
    DequeCollection.prototype._Erase_by_range = function (first, last) {
        this._Notify_erase(first, last);
        return _super.prototype._Erase_by_range.call(this, first, last);
    };
    /* ---------------------------------------------------------
        REFRESH
    --------------------------------------------------------- */
    /**
     * @inheritdoc
     */
    DequeCollection.prototype.set = function (index, val) {
        _super.prototype.set.call(this, index, val);
        this.refresh(this.begin().advance(index));
    };
    DequeCollection.prototype.refresh = function (first, last) {
        if (first === void 0) { first = null; }
        if (last === void 0) { last = null; }
        if (first === null) {
            first = this.begin();
            last = this.end();
        }
        else if (last === null)
            last = first.next();
        this.dispatchEvent(new CollectionEvent_1.CollectionEvent("refresh", first, last));
    };
    /* =========================================================
        EVENT DISPATCHER
            - NOTIFIERS
            - ACCESSORS
    ============================================================
        NOTIFIERS
    --------------------------------------------------------- */
    /**
     * @inheritdoc
     */
    DequeCollection.prototype.dispatchEvent = function (event) {
        if (this.dispatcher_)
            this.dispatcher_.dispatchEvent(event);
    };
    /**
     * @hidden
     */
    DequeCollection.prototype._Notify_insert = function (first, last) {
        this.dispatchEvent(new CollectionEvent_1.CollectionEvent("insert", first, last));
    };
    /**
     * @hidden
     */
    DequeCollection.prototype._Notify_erase = function (first, last) {
        this.dispatchEvent(new CollectionEvent_1.CollectionEvent("erase", first, last));
    };
    /* ---------------------------------------------------------
        ACCESSORS
    --------------------------------------------------------- */
    /**
     * @inheritdoc
     */
    DequeCollection.prototype.hasEventListener = function (type) {
        return this.dispatcher_.hasEventListener(type);
    };
    /**
     * @inheritdoc
     */
    DequeCollection.prototype.addEventListener = function (type, listener) {
        this.dispatcher_.addEventListener(type, listener);
    };
    /**
     * @inheritdoc
     */
    DequeCollection.prototype.removeEventListener = function (type, listener) {
        this.dispatcher_.removeEventListener(type, listener);
    };
    return DequeCollection;
}(Deque_1.Deque));
exports.DequeCollection = DequeCollection;
(function (DequeCollection) {
    DequeCollection.Event = CollectionEvent_1.CollectionEvent;
    DequeCollection.Iterator = Deque_1.Deque.Iterator;
    DequeCollection.ReverseIterator = Deque_1.Deque.ReverseIterator;
})(DequeCollection = exports.DequeCollection || (exports.DequeCollection = {}));
exports.DequeCollection = DequeCollection;
var old_swap = Deque_1.Deque.prototype.swap;
Deque_1.Deque.prototype.swap = function (obj) {
    old_swap.call(this, obj);
    if (this instanceof DequeCollection)
        this.refresh();
    if (obj instanceof DequeCollection)
        obj.refresh();
};
