// ecol/lib/maps/HashMultiMapCollection.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var HashMultiMap_1 = require("tstl/container/HashMultiMap");
var CollectionEvent_1 = require("../basic/CollectionEvent");
var EventDispatcher_1 = require("../basic/EventDispatcher");
require("./internal");
var HashMultiMapCollection = /** @class */ (function (_super) {
    __extends(HashMultiMapCollection, _super);
    function HashMultiMapCollection() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * @hidden
         */
        _this.dispatcher_ = new EventDispatcher_1.EventDispatcher();
        return _this;
    }
    /* ---------------------------------------------------------
        CONSTRUCTORS
    --------------------------------------------------------- */
    // using super.constructor;
    HashMultiMapCollection.prototype.clear = function () {
        var first = this.begin();
        var last = this.end();
        _super.prototype.clear.call(this);
        this.dispatchEvent(new CollectionEvent_1.CollectionEvent("erase", first, last));
    };
    /* ---------------------------------------------------------
        ELEMENTS I/O
    --------------------------------------------------------- */
    /**
     * @hidden
     */
    HashMultiMapCollection.prototype._Handle_insert = function (first, last) {
        _super.prototype._Handle_insert.call(this, first, last);
        this.dispatchEvent(new CollectionEvent_1.CollectionEvent("insert", first, last));
    };
    /**
     * @hidden
     */
    HashMultiMapCollection.prototype._Handle_erase = function (first, last) {
        this._Handle_erase(first, last);
        this.dispatchEvent(new CollectionEvent_1.CollectionEvent("erase", first, last));
    };
    /* =========================================================
        EVENT DISPATCHER
            - NOTIFIERS
            - ACCESSORS
    ============================================================
        NOTIFIERS
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    HashMultiMapCollection.prototype.dispatchEvent = function (event) {
        if (this.dispatcher_)
            this.dispatcher_.dispatchEvent(event);
    };
    HashMultiMapCollection.prototype.refresh = function (first, last) {
        if (first === void 0) { first = null; }
        if (last === void 0) { last = null; }
        if (first === null) {
            first = this.begin();
            last = this.end();
        }
        else if (last === null)
            last = first.next();
        this.dispatchEvent(new CollectionEvent_1.CollectionEvent("refresh", first, last));
    };
    /* ---------------------------------------------------------
        ACCESSORS
    -------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    HashMultiMapCollection.prototype.hasEventListener = function (type) {
        return this.dispatcher_.hasEventListener(type);
    };
    /**
     * @inheritDoc
     */
    HashMultiMapCollection.prototype.addEventListener = function (type, listener) {
        this.dispatcher_.addEventListener(type, listener);
    };
    /**
     * @inheritDoc
     */
    HashMultiMapCollection.prototype.removeEventListener = function (type, listener) {
        this.dispatcher_.removeEventListener(type, listener);
    };
    return HashMultiMapCollection;
}(HashMultiMap_1.HashMultiMap));
exports.HashMultiMapCollection = HashMultiMapCollection;
(function (HashMultiMapCollection) {
    HashMultiMapCollection.Event = CollectionEvent_1.CollectionEvent;
    HashMultiMapCollection.Iterator = HashMultiMap_1.HashMultiMap.Iterator;
    HashMultiMapCollection.ReverseIterator = HashMultiMap_1.HashMultiMap.ReverseIterator;
})(HashMultiMapCollection = exports.HashMultiMapCollection || (exports.HashMultiMapCollection = {}));
exports.HashMultiMapCollection = HashMultiMapCollection;
var old_swap = HashMultiMap_1.HashMultiMap.prototype.swap;
HashMultiMap_1.HashMultiMap.prototype.swap = function (obj) {
    old_swap.call(this, obj);
    if (this instanceof HashMultiMapCollection)
        this.refresh();
    if (obj instanceof HashMultiMapCollection)
        obj.refresh();
};
