// samchon/protocol/entity/EntityCollection.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var collections = require("ecol");
var IEntityGroup_1 = require("./IEntityGroup");
/**
 * @inheritdoc
 */
var EntityArrayCollection = /** @class */ (function (_super) {
    __extends(EntityArrayCollection, _super);
    function EntityArrayCollection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /* ------------------------------------------------------------------
        CONSTRUCTORS
    ------------------------------------------------------------------ */
    // using super::super;
    /**
     * @inheritdoc
     */
    EntityArrayCollection.prototype.construct = function (xml) {
        IEntityGroup_1.IEntityGroup.construct(this, xml);
    };
    /* ------------------------------------------------------------------
        GETTERS
    ------------------------------------------------------------------ */
    /**
     * @inheritdoc
     */
    EntityArrayCollection.prototype.key = function () {
        return "";
    };
    /**
     * @inheritdoc
     */
    EntityArrayCollection.prototype.has = function (key) {
        return IEntityGroup_1.IEntityGroup.has(this, key);
    };
    /**
     * @inheritdoc
     */
    EntityArrayCollection.prototype.count = function (key) {
        return IEntityGroup_1.IEntityGroup.count(this, key);
    };
    /**
     * @inheritdoc
     */
    EntityArrayCollection.prototype.get = function (key) {
        return IEntityGroup_1.IEntityGroup.get(this, key);
    };
    /**
     * @inheritdoc
     */
    EntityArrayCollection.prototype.toXML = function () {
        return IEntityGroup_1.IEntityGroup.toXML(this);
    };
    return EntityArrayCollection;
}(collections.ArrayCollection));
exports.EntityArrayCollection = EntityArrayCollection;
/**
 * @inheritdoc
 */
var EntityListCollection = /** @class */ (function (_super) {
    __extends(EntityListCollection, _super);
    function EntityListCollection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /* ------------------------------------------------------------------
        CONSTRUCTORS
    ------------------------------------------------------------------ */
    // using super::super;
    /**
     * @inheritdoc
     */
    EntityListCollection.prototype.construct = function (xml) {
        IEntityGroup_1.IEntityGroup.construct(this, xml);
    };
    /* ------------------------------------------------------------------
        GETTERS
    ------------------------------------------------------------------ */
    /**
     * @inheritdoc
     */
    EntityListCollection.prototype.key = function () {
        return "";
    };
    /**
     * @inheritdoc
     */
    EntityListCollection.prototype.has = function (key) {
        return IEntityGroup_1.IEntityGroup.has(this, key);
    };
    /**
     * @inheritdoc
     */
    EntityListCollection.prototype.count = function (key) {
        return IEntityGroup_1.IEntityGroup.count(this, key);
    };
    /**
     * @inheritdoc
     */
    EntityListCollection.prototype.get = function (key) {
        return IEntityGroup_1.IEntityGroup.get(this, key);
    };
    /**
     * @inheritdoc
     */
    EntityListCollection.prototype.toXML = function () {
        return IEntityGroup_1.IEntityGroup.toXML(this);
    };
    return EntityListCollection;
}(collections.ListCollection));
exports.EntityListCollection = EntityListCollection;
/**
 * @inheritdoc
 */
var EntityDequeCollection = /** @class */ (function (_super) {
    __extends(EntityDequeCollection, _super);
    function EntityDequeCollection() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /* ------------------------------------------------------------------
        CONSTRUCTORS
    ------------------------------------------------------------------ */
    // using super::super;
    /**
     * @inheritdoc
     */
    EntityDequeCollection.prototype.construct = function (xml) {
        IEntityGroup_1.IEntityGroup.construct(this, xml);
    };
    /* ------------------------------------------------------------------
        GETTERS
    ------------------------------------------------------------------ */
    /**
     * @inheritdoc
     */
    EntityDequeCollection.prototype.key = function () {
        return "";
    };
    /**
     * @inheritdoc
     */
    EntityDequeCollection.prototype.has = function (key) {
        return IEntityGroup_1.IEntityGroup.has(this, key);
    };
    /**
     * @inheritdoc
     */
    EntityDequeCollection.prototype.count = function (key) {
        return IEntityGroup_1.IEntityGroup.count(this, key);
    };
    /**
     * @inheritdoc
     */
    EntityDequeCollection.prototype.get = function (key) {
        return IEntityGroup_1.IEntityGroup.get(this, key);
    };
    /**
     * @inheritdoc
     */
    EntityDequeCollection.prototype.toXML = function () {
        return IEntityGroup_1.IEntityGroup.toXML(this);
    };
    return EntityDequeCollection;
}(collections.DequeCollection));
exports.EntityDequeCollection = EntityDequeCollection;
