// tstl/thread/Mutex.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var Queue_1 = require("../container/Queue");
var RuntimeError_1 = require("../exception/RuntimeError");
/**
 * Mutex.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var Mutex = /** @class */ (function () {
    /* ---------------------------------------------------------
        CONSTRUCTORS
    --------------------------------------------------------- */
    /**
     * Default Constructor.
     */
    function Mutex() {
        this.lock_count_ = 0;
        this.resolvers_ = new Queue_1.Queue();
    }
    /* ---------------------------------------------------------
        LOCK & UNLOCK
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    Mutex.prototype.lock = function () {
        var _this = this;
        return new Promise(function (resolve) {
            if (_this.lock_count_++ === 0)
                resolve();
            else
                _this.resolvers_.push(resolve);
        });
    };
    /**
     * @inheritDoc
     */
    Mutex.prototype.try_lock = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (this.lock_count_ !== 0)
                    return [2 /*return*/, false]; // HAVE LOCKED
                ++this.lock_count_;
                return [2 /*return*/, true];
            });
        });
    };
    /**
     * @inheritDoc
     */
    Mutex.prototype.unlock = function () {
        return __awaiter(this, void 0, void 0, function () {
            var fn;
            return __generator(this, function (_a) {
                if (this.lock_count_ === 0)
                    throw new RuntimeError_1.RangeError("This mutex is free.");
                --this.lock_count_; // DECREASE LOCKED COUNT
                if (this.resolvers_.empty() === false) {
                    fn = this.resolvers_.front();
                    this.resolvers_.pop(); // POP FIRST
                    fn(); // AND CALL LATER
                }
                return [2 /*return*/];
            });
        });
    };
    return Mutex;
}());
exports.Mutex = Mutex;
exports.mutex = Mutex;
